redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java

```diff
@@ -202,18 +227,38 @@ public boolean exceedsInProgressLimit(Task task) {
 		if(!taskDefinition.isPresent()) {
 			return false;
 		}
+		
 		int limit = taskDefinition.get().concurrencyLimit();
-		if(limit <= 0) {
-			return false;
-		}
-
 		long current = getInProgressTaskCount(task.getTaskDefName());
-		if(current >= limit) {
+		if(limit > 0 && current >= limit) {
 			logger.info("Task execution count limited. task - {}:{}, limit: {}, current: {}", task.getTaskId(), task.getTaskDefName(), limit, current);
 			Monitors.recordTaskConcurrentExecutionLimited(task.getTaskDefName(), limit);
 			return true;
 		}
 
+		if(enableGlobalTaskConcurrentExecLimit || enableLocalTaskConcurrentExecLimit) {
+			logger.debug("ExecLimit acquire processing {} with id {}", task.getReferenceTaskName(), task.getTaskId());
+			WorkflowTask wfTask = task.getWorkflowTask();
+			if(enableGlobalTaskConcurrentExecLimit && wfTask.getGlobalConcurrentExecutionLimit() > 0) {
+				logger.debug("acquiring semaphore " + executionLimitingSemaphoreName + task.getReferenceTaskName());
+				if(!semaphoreDAO.tryAcquire(executionLimitingSemaphoreName + task.getReferenceTaskName(), task.getTaskId(), wfTask.getGlobalConcurrentExecutionLimit(), taskDefinition.get().getTimeoutSeconds() * 1000.0)){
```

Comment:
Since the semaphore uses task reference names to constrain concurrency limits, it breaks the isolation construct among independent tasks. Since the same task reference name can be used by different tasks in different workflows, setting the global limit for one would impact the execution of an unrelated task in an unrelated workflow, which is an unwanted side-effect.

Reply:
I understand your concern here but l'm aware and that's exactly what I want to happened. In our system all the `taskReferenceName` s are controlled by another microservice and that's how we can guarantee zero conflicts. This change won't impact any currently defined workflows, not in our system, nor in anyone else's system because the default value for the limits is 0 and when it's 0 my code is basically skipped. The idea here is that if you are familiar with all the taskReferenceName in your system, you could enable this feature if you need to.

Reply:
> In our system all the taskReferenceName s are controlled by another microservice and that's how we can guarantee zero conflicts.
The idea here is that if you are familiar with all the taskReferenceName in your system, you could enable this feature if you need to.

This is more of an exception than the norm. 

The contract that Conductor requires from tasks is to have unique names across the system and for workflow tasks to have unique reference names within a given workflow. However, there is no restriction for workflow tasks to be unique across workflows. 

> because the default value for the limits is 0 and when it's 0 my code is basically skipped.

While this is true, this feature now becomes unusable to any other user because it is not a constraint that is enforced by Conductor. Also, if a user were to start using this feature, it would unknowingly cause side-effects in their other workflows because unlike the task name validation constraint or the unique task reference name validation constraint, there is no validation preventing users from re-using the same task reference name across workflows.

Reply:
@apanicker-nflx
I understand. 
What do you think about adding another string field to use as the semaphore name instead of the `taskReferenceName`? Say `concurrentLockName` or `concurrentSemaphoreName`?

Reply:
@TwoUnderscorez Adding a new field for this purpose does not completely solve the isolation/side-effect argument mentioned above. Workflow owners for independent workflows could still use the same name within these params leading to issues.

Based on our internal discussion, we think it would be best to extract the ConcurrentExecutionLimits into a separate interface with the currently provided implementation as the default. The solution in this PR could be added as a pluggable implementation of this interface for users looking to solve concurrency limiting at a workflow task level. The caveat here being the external system that helps to keep task reference names unique across workflows.

Please let us know if you would be willing to undertake this interface extraction.

Reply:
@apanicker-nflx I see.
So have `exceedsInProgressLiint` in all the execution DAOs call on that interface? The only problem with that is I'll run into the same problem as here #1951 because I want to add the `taskReferenceName` based looking - not replace it with the existing one. Basically call all the concurrent limiting implementations in an order specified by the user and if one returns false, stop and return false.

Reply:
> So have exceedsInProgressLiint in all the execution DAOs call on that interface? 

Yes, that interface would then provide the implementation for the `exceedsInProgressLimit`. 

> I'll run into the same problem as here #1951 because I want to add the taskReferenceName based looking - not replace it with the existing one.

Here again, you could add this as an extension of what is currently provided rather than as a replacement, similar to the status listener use case.

Reply:
@apanicker-nflx Alright. So the `exceedsInProgressLiint` will do the task def limit check, then call the new interface, then rate limit check? 
Do you think we should open a issue to discuss this, because I have a few more questions like:
With this PR, I'll add my implementation of the interface, correct?

If so, I'll need a semaphore, is the DAO fine? Is the implementation fine?

I don't think that the task def concurrent execution limit implementation is not as reliable as I would like it to be if I were using it. Should I switch it to use a semaphore? (Probably a different PR though)

Could you please provide me the definition you were thinking about for that interface? Just to make sure we don't have a misunderstanding of some minor detail that ends up not being minor at all.

Reply:
That is a great idea, let's continue this discussion in a new issue rather than as comments in this PR. It would provide visibility to the problem you are trying to solve to the community and we can gather more inputs. Thanks.

Reply:
@apanicker-nflx I've created issue #2066 .